[WIP] Segwit4

sipa opened the pull request.

sipa committed:
Add segregated witness transaction serialization

Contains refactorings by Eric Lombrozo.
Contains fixup by Nicolas Dorier.

sipa committed:
Witness commitment validation

Includes a fix by Suhas Daftuar

sipa committed:
Script validation logic for witnesses

sipa committed:
Introduce and preferentially peer with NODE_WITNESS service bit

Service bit logic by Nicolas Dorier.

sipa committed:
Enable SCRIPT_VERIFY_WITNESS for mempool transactions

sipa committed:
Activate script consensus rules through BIP9

sipa committed:
Signing tests

sipa committed:
Add signature version 1 with updated sighash

Includes simplifications by Eric Lombrozo.

sipa committed:
Only download blocks from witness peers after fork

sipa committed:
Add witness address RPCs (using P2SH)

Includes support for pushkeyhash wit v0 by Alex Morcos.

sipa committed:
Create segnet4

morcos committed:
Add command line options to loosen mempool acceptance rules

morcos committed:
Add rpc test for segwit

Amended by Pieter Wuille to use multisig 1-of-1 for P2WSH tests, and BIP9
based switchover logic.

sipa committed:
Implement block size/sigop cost rules, limits, and GBT support

Includes changes by Suhas Daftuar and Luke-jr.

CodeShark committed:
Removed ppszTypeName from protocol.cpp

theuni committed:
qt: Work (don't crash) with -segnet

CodeShark committed:
getdata enum issue fix

sdaftuar committed:
Increase MAX_PROTOCOL_MESSAGE_LENGTH

Witness blocks can be greater than 2MiB, but cannot be validly greater
than 4MB.

sipa committed:
Add segnet seed nodes

jl2012 committed:
Return witness data

Includes RPC field name changes by Luke-jr.

afk11 committed:
bitcoinconsensus: add method that accepts amount, and return error when verify_script receives VERIFY_WITNESS flag

script_tests: always test bitcoinconsensus_verify_script_with_amount if VERIFY_WITNESS isn't set

Rename internal method + make it static

trim bitcoinconsensus_ prefix

Add SERIALIZE_TRANSACTION_WITNESS flag

sipa committed:
Add rewind logic to deal with post-fork software updates

mruddy commented:
@sipa minor but the first thing i tried was bitcoin-qt and it crashed. this fixes that crash:
```
src/qt/networkstyle.cpp
-    {"segnet", QAPP_APP_NAME_SEGNET, 30, 30, QT_TRANSLATE_NOOP("SplashScreen", "[segnet]")},
+    {"segnet4", QAPP_APP_NAME_SEGNET, 30, 30, QT_TRANSLATE_NOOP("SplashScreen", "[segnet4]")},
```

sipa commented:
Thanks!

sipa committed:
--- [SEGWIT] begin: segnet ---

sipa committed:
--- [SEGWIT] begin: P2P/node/consensus ---

sipa committed:
--- [SEGWIT] begin: wallet ---

sipa committed:
--- [SEGWIT] begin: tests ---

sipa committed:
Observe input amounts: verification

sipa committed:
Witness script signing

sdaftuar commented on src/main.cpp:
```diff
@@ -2299,6 +2299,11 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     }
 
+    // Start enforcing WITNESS rules using versionbits logic.
+    if (IsWitnessEnabled(pindex->pprev, chainparams.GetConsensus())) {
```
Just a reminder to update the commit message for this commit, to reflect versionbits usage (rather than v5 blocks).

sipa replied:
Fixed.

sdaftuar commented on src/chainparams.cpp:
```diff
@@ -92,6 +93,11 @@ class CMainParams : public CChainParams {
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].bit = 1;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nStartTime = 2000000000; // Far in the future
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 2100000000;
+
```
These should say DEPLOYMENT_WITNESS

sipa replied:
Fixed.

sdaftuar commented on src/chainparams.cpp:
```diff
@@ -185,6 +191,11 @@ class CTestNetParams : public CChainParams {
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].bit = 1;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nStartTime = 2000000000; // Far in the future
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 2100000000;
+
```
DEPLOYMENT_WITNESS again

sipa replied:
Fixed.

sdaftuar commented on src/main.cpp:
```diff
@@ -5911,10 +6135,12 @@ bool SendMessages(CNode* pto)
-                MarkBlockAsInFlight(pto->GetId(), pindex->GetBlockHash(), consensusParams, pindex);
-                LogPrint("net", "Requesting block %s (%d) peer=%d\n", pindex->GetBlockHash().ToString(),
-                    pindex->nHeight, pto->id);
+                if (State(pto->GetId())->fHaveWitness || !IsWitnessEnabled(pindex, consensusParams)) {
```
If I am understanding this right, then I think we wouldn't download (via parallel-fetch at least) the very last block before segwit is activated from a non-NODE_WITNESS peer?  I can't imagine this being a significant issue but I thought it might be clearer anyway to make this `!IsWitnessEnabled(pindex->pprev, )`.

sipa replied:
Fixed.

sdaftuar commented on src/main.cpp:
```diff
@@ -5167,7 +5390,8 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 if (!(pindexWalk->nStatus & BLOCK_HAVE_DATA) &&
-                        !mapBlocksInFlight.count(pindexWalk->GetBlockHash())) {
+                        !mapBlocksInFlight.count(pindexWalk->GetBlockHash()) &&
+                        (!IsWitnessEnabled(chainActive.Tip(), chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness)) {
```
I think this should be `!IsWitnessEnabled(pindexWalk->pprev, ...)`, as otherwise if you were trying to download the last couple of blocks that would end with segwit activation, you could accidentally download the non-witness versions of those blocks.

sipa replied:
Yes, much better!

NicolasDorier committed:
Add transaction tests for segwit

...with the four types of segwit payment, as well as all sighash combinaisons.

sipa committed:
Add segwit support to script_tests

sdaftuar commented on src/net.cpp:
```diff
@@ -92,9 +92,9 @@ std::string strSubVersion;
-CCriticalSection cs_mapRelay;
+map<uint256, CTransaction> mapRelayTx;
+deque<pair<int64_t, uint256> > vRelayTxExpiration;
+CCriticalSection cs_mapRelayTx;
```
Perhaps we should we also change `mapAlreadyAskedFor` to be a `limitedmap<uint256, int64_t>`?  In the current implementation, we could have 2 requests in flight for the same txid, one for the witness version and one for the non-witness version, if we an upgraded- and non-upgraded-node both inv'ed the same transaction.

sipa committed:
Autogeneration support for witness in script_tests

sdaftuar commented on src/main.cpp:
```diff
@@ -1356,10 +1379,21 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState &state, const C
-        if (!CheckInputs(tx, state, view, true, STANDARD_SCRIPT_VERIFY_FLAGS, true))
+        if (!CheckInputs(tx, state, view, true, scriptVerifyFlags, true)) {
+            if (CheckInputs(tx, state, view, true, scriptVerifyFlags & ~(SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_CLEANSTACK), true) &&
+                !CheckInputs(tx, state, view, true, scriptVerifyFlags & ~SCRIPT_VERIFY_CLEANSTACK, true)) {
```
Perhaps add a comment  reminding the reader that unsetting SCRIPT_VERIFY_CLEANSTACK here doesn't turn off the requirement that CLEANSTACK is always enforced for witness validation?  (I thought there was a bug here until I realized that.)

sdaftuar committed:
Update p2p test framework with segwit support

mininode now supports witness transactions/blocks, blocktools
has a helper for adding witness commitments to blocks, and script
has a function to calculate hashes for signature under sigversion
1, used by segwit.

sdaftuar committed:
P2P test for segwit

NicolasDorier committed:
signrawtransaction can sign P2WSH

sdaftuar commented on src/main.cpp:
```diff
@@ -2676,6 +2676,9 @@ bool static DisconnectTip(CValidationState& state, const Consensus::Params& cons
     if (!FlushStateToDisk(state, FLUSH_STATE_IF_NEEDED))
         return false;
+    if (fBare)
+        return true;
```
I don't think we can return true here without calling `UpdateTip`?  Otherwise chainActive is not updated.

sdaftuar commented on src/main.cpp:
```diff
@@ -4070,6 +4073,53 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+        pindex->nUndoPos = 0;
+        // Make sure it gets written.
+        setDirtyBlockIndex.insert(pindex);
+    }
```
Need to add `pindex = pindex->pprev`, to avoid infinite loop.

sdaftuar commented on src/main.cpp:
```diff
@@ -4070,6 +4073,53 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+        return false;
+    }
+
+    return true;
```
I think we need to update setBlockIndexCandidates before returning from this function; with the above two fixes I'm now seeing an assertion failure in CheckBlockIndex.

sdaftuar commented on src/main.cpp:
```diff
@@ -4070,6 +4073,53 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    // to disk before writing the chainstate, resulting in a failure to continue if interrupted.
+    while (pindex->nHeight >= nHeight) {
+        // Reduce validity
+        pindex->nStatus = std::min<unsigned int>(pindex->nStatus & BLOCK_VALID_MASK, BLOCK_VALID_TREE) | (pindex->nStatus & ~BLOCK_VALID_MASK);
```
To be consistent with the assumptions tested in `CheckBlockIndex`, I think we also need to set:
```
pindex->nTx = 0;
pindex->nChainTx = 0;
pindex->nSequenceId = 0;
setBlockIndexCandidates.erase(pindex);
```

sdaftuar commented on src/main.cpp:
```diff
@@ -4070,6 +4073,53 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    CValidationState state;
+    CBlockIndex* pindex = chainActive.Tip();
+    while (chainActive.Height() >= nHeight) {
+        if (!DisconnectTip(state, Params().GetConsensus(), true)) {
```
To better support pruning nodes that might upgrade post-activation, I suggest we add a check  before this call to `DisconnectTip()`, something like:
```
if (fHavePruned && !(chainActive.Tip() & BLOCK_HAVE_DATA)) {
    // graceful error telling user to reindex and redownload the whole block chain to upgrade post-segwit activation
}
```

sdaftuar commented on src/main.cpp:
```diff
@@ -4073,6 +4076,69 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+            setDirtyBlockIndex.insert(pindexIter);
+            // Update setBlockIndexCandidates
+            setBlockIndexCandidates.erase(pindexIter);
+            if ((!IsWitnessEnabled(pindexIter->pprev->pprev, params) || (pindexIter->pprev->nStatus & BLOCK_OPT_WITNESS)) && pindexIter->pprev->IsValid(BLOCK_VALID_TRANSACTIONS) && pindexIter->pprev->nChainTx) {
```
Looking at `pprev` here is confusing, since you need to repeat the condition in the `if()` above to make this work.  I think this is right, but perhaps it'd be easier to read if, in this loop, we just consider whether to add the pindexIter we're looking at to setBlockIndexCandidates, say by rewriting this as follows:
```
    ....
    setBlockIndexCandidates.erase(pindexIter);
} else if (pindexIter->IsValid(BLOCK_VALID_TRANSACTIONS) && pindexIter->nChainTx) {
    setBlockIndexCandidates.insert(pindexIter);
}
```
This is inefficient from the perspective of adding more entries than is strictly needed to `setBlockIndexCandidates`, and therefore the prune below will take a little longer, but that seems like a small effect?

sdaftuar replied:
Actually I take back what I said about being inefficient -- I think this logic is wrong as written, but would be correct in my reformulation.

sipa replied:
Agree, after talking elsewhere. I'll switch to your version.

sipa committed:
Don't check the genesis block PoW

Inspired by Jorge Timon's "Testchain: Don't check the genesis block", but
restricted to just PoW check.

Code by Matt Corallo, originally submitted as a patch to segwit.

mruddy commented on src/script/interpreter.cpp:
```diff
@@ -1106,8 +1110,64 @@ class CTransactionSignatureSerializer {
+            hashPrevouts = ss.GetHash(); // TODO: cache this value for all signatures in a transaction
+        }
+
+        if (!(nHashType & SIGHASH_ANYONECANPAY) && (nHashType & 0x1f) != SIGHASH_SINGLE && (nHashType & 0x1f) != SIGHASH_NONE) {
```
Besides trying to make this version of the sig hash algorithm act similar to the original one, is there a reason why SINGLE and NONE are involved in this condition? To me SINGLE and NONE are flags controlling which output data gets committed to and it seems awkward that they also control which input sequences get committed. Just wondering if it makes sense to stop considering SINGLE and NONE here and then merge this loop with the previous one?

jl2012 replied:
I think the original intention is to allow replacement by nsequence if one of these types is used

mruddy replied:
Right, I had written that off as vestigial since I don't see standard mempool policy being changed to accept non-final transactions.
As a signer, why would switching from caring about all outputs to caring only about a subset of outputs make me not care about other input sequence numbers?
When is that useful?
The reason for keeping this would have to be for some offline protocol. But, I can't think of one where this actually works out being necessary.
Also, this coupled with BIP68 allows a counterparty to change the time at which their signature becomes usable (i.e.- when the transaction becomes BIP68 final and relayable).
This seems benign, but if I care about other inputs, why would I not care about this effect?

jl2012 replied:
If you don't care all outputs, other people may want to update their outputs, for example, with opt-in RBF (or to disable opt-in RBF to make the tx final)

If you don't own all private keys of all inputs, other people already have the ability to invalidate their signature through double spending. So BIP68 makes no big difference here.

jl2012 replied:
Also for BIP68: without signing nSequence of other inputs, other people may: 1. reduce the locktime to make an invalid signature valid; or 2. increase the locktime to make a valid signature invalid.

In the first case, you could just set your own nSequence to make sure the tx won't be mined until your preset time. So you have the full control.

In the second case, since the original signature is still valid, it could be confirmed immediately. So nothing changed

mruddy replied:
> If you don't care all outputs, other people may want to update their outputs, for example, with opt-in RBF (or to disable opt-in RBF to make the tx final)

RBF only matters once the tx is propagated and the tx has to be "final" and "BIP68 final" before it'll propagate.
Before it's broadcast, they can freely change their outputs since a SINGLE or NONE signature allows that.
Once a tx gets propagated that opts-out of RBF, then they wouldn't be able to change their sequence to opt-in and be able to get the new tx relayed.

Since I care about all the inputs, I think I'd want to have to sign them changing to finalize early if that's different than what I was expecting.
If I anticipated them using a final sequence number (because that's what the protocol defined or what the parties agreed upon), then I'd just sign that to begin with.

Same with the BIP68 stuff. It seems mostly benign (the first case you mentioned is a potential stumbling point that a protocol might have to account for), but I haven't seen a compelling reason to give other parties the flexibility of changing sequence numbers when I care about all the inputs.

I'm thinking that the proper sequence number to sign would be defined by the protocol or coordinated based on the parties' preferences for locktime enablement and replaceability. I think that's reasonable, but maybe I'm wrong. Eh, it probably doesn't matter. I was just curious.
